File: Data Structures & Algorithms/subsets-ii/submission-1.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []

        def backtrack(i: int, cur: List[int]):
            if i >= len(nums):
                res.append(cur[:])
                return

            cur.append(nums[i])
            backtrack(i + 1, cur)
            cur.pop()
            while i + 1 < len(nums) and nums[i] == nums[i + 1]:
                i += 1
            backtrack(i + 1, cur)

        backtrack(0, [])
        return res

File: Data Structures & Algorithms/permutations/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    permute(nums) {
        let res = []
        let cur = []

        function backtrack(pick) {
            if (cur.length === nums.length) {
                res.push([...cur])
                return
            }

            for (let num of nums) {
                if (!pick[num]) {
                    cur.push(num)
                    pick[num] = true

                    backtrack(pick)
                    
                    cur.pop()
                    pick[num] = false
                }
            }
        }

        backtrack([])

        return res
    }
}
